### ISSUE TITLE
Server isn't asynchronous

### ISSUE THREAD

**Issue by ghost** [little-dude/rmp-rpc#19]
I'm trying to make it possible for multiple clients to call RPC functions in parallel. I took the example `server.rs` and tweaked it a bit, using a five-second timer to simulate a long computation. https://pastebin.com/tENjUKqt

However, if I run two clients, they make requests synchronously, with the first one blocking the second one. Is it possible for each request to run at the same time?

**Comment by little-dude** [comment]
The code looks ok to me... Do you mind sharing the client code?

**Comment by ghost** [comment]
```python
from mprpc import RPCClient
client = RPCClient('127.0.0.1', 54321)
print(client.call("echo", "hello"))
```

**Comment by little-dude** [comment]
Thanks. Seems like a major bug indeed... I'm so sorry about that. I'll take a longer look this week end.

**Comment by little-dude** [comment]
I actually don't really know what's going on... The `InnerEndpoint` is not getting polled despite tokio getting a read event when the second request arrives:

```
TRACE 2018-10-25T19:40:11Z: tokio_reactor: event Readable Token(0)        
DEBUG 2018-10-25T19:40:11Z: tokio_reactor: loop process - 1 events, 0.000s
```

**Comment by ghost** [comment]
Welp.

**Comment by little-dude** [comment]
@Riamse I only had a quick look. I'll debug further this week end.

**Comment by little-dude** [comment]
For a given connection, things work fine. It's really for multiple connections that things for wrong. For instance with [messagepack-rpc-python](https://github.com/msgpack-rpc/msgpack-rpc-python):

```python
import time                                                          
import msgpackrpc

client = client = msgpackrpc.Client(msgpackrpc.Address("127.0.0.1", 54321))
start = time.time()
requests = []
for i in range(0, 1000):
    requests.append(client.call_async('hello', 'hello'))
for req in requests:
    req.get()
end = time.time()
print(end - start)
```

Prints `5.02` for me.

**Comment by little-dude** [comment]
More debugging: I'm running the following python script:

```python
# issue19.py
import os
import time
import mprpc


pid = os.getpid()
print(f'{pid}: {time.time()}: client connecting')
client = mprpc.RPCClient('127.0.0.1', 54321)
print(f'{pid}: {time.time()}: client connected, sending request')
client.call('do_long_computation', 5)
print(f'{pid}: {time.time()}: got response')
```

With: `python issue19.py & ; sleep 2 ; python issue19.py`.

What should happen:
- `t0`: the first client connects and sends its request
- `t0+2s`: the second client connects and sends its request
- `t0+5s`: the first request is handled and the response is sent back
- `t0+7s`: the second request is handled and the response is sent back

What actually happens:

- `t0`: the first client connects and sends its request
- `t0+2s`: the second client connects but `rmp-rpc` does not process the connection
- `t0+5s`: the first request is handled and the response is sent back. Finally, we handle the second connection, and start processing the second request
- `t0+10s`: the second request is handled, and the response is sent back

Below are the full logs

--------------------------------------

Here is the server starting
```
TRACE 22:09:08: mio::poll: registering with poller
TRACE 22:09:08: tokio_threadpool::builder: build; num-workers=4
DEBUG 22:09:08: tokio_reactor::background: starting background reactor
TRACE 22:09:08: tokio_threadpool::sender: execute; count=1
TRACE 22:09:08: tokio_threadpool::pool: submit to existing worker; idx=3; state=worker::State { lifecycle: Shutdown, is_pushed: true }
TRACE 22:09:08: tokio_threadpool::pool: shutdown; state=pool::State { lifecycle: Running, num_futures: 1 }
TRACE 22:09:08: tokio_threadpool::pool:   -> transitioned to shutdown
TRACE 22:09:08: tokio_threadpool::worker: found work while draining; signal_work
TRACE 22:09:08: tokio_threadpool::pool: signal_work -- spawn; idx=2
TRACE 22:09:08: tokio_threadpool::task: Task::run; state=Running
TRACE 22:09:08: mio::poll: registering with poller
TRACE 22:09:08: tokio_threadpool::task:     -> not ready
TRACE 22:09:08: tokio_threadpool::worker: Worker::sleep; worker=WorkerId(3)
TRACE 22:09:08: tokio_threadpool::worker: Worker::sleep; worker=WorkerId(2)
TRACE 22:09:08: tokio_threadpool::worker:   sleeping -- push to stack; idx=3
TRACE 22:09:08: tokio_threadpool::worker:   sleeping -- push to stack; idx=2
TRACE 22:09:08: tokio_threadpool::worker:     -> starting to sleep; idx=2
TRACE 22:09:08: tokio_threadpool::worker:     -> starting to sleep; idx=3
```

`t0`: The first client connects and sends its request right away

```
TRACE 22:09:12: tokio_reactor: event Readable Token(0)  <---------------------------- The client connects
TRACE 22:09:12: tokio_threadpool::notifier: Notifier::notify; id=0x7f0387652080
TRACE 22:09:12: tokio_threadpool::pool: submit to existing worker; idx=2; state=worker::State { lifecycle: Sleeping, is_pushed: true }
DEBUG 22:09:12: tokio_reactor: loop process - 1 events, 0.000s
TRACE 22:09:12: tokio_threadpool::worker:     -> wakeup; idx=2
TRACE 22:09:12: tokio_threadpool::worker: found work while draining; signal_work
TRACE 22:09:12: tokio_threadpool::pool: signal_work -- wakeup; idx=3
TRACE 22:09:12: tokio_threadpool::task: Task::run; state=Running
TRACE 22:09:12: tokio_threadpool::worker:     -> wakeup; idx=3
TRACE 22:09:12: tokio_threadpool::worker: Worker::sleep; worker=WorkerId(3)
TRACE 22:09:12: tokio_threadpool::worker:   sleeping -- push to stack; idx=3
 INFO 22:09:12: server: new connection TcpStream { addr: V4(127.0.0.1:54321), peer: V4(127.0.0.1:44694), fd: 7 }
TRACE 22:09:12: tokio_threadpool::worker:     -> starting to sleep; idx=3
 INFO 22:09:12: server: spawning a new Server <--------------------------------------- We spawn a new endpoint on the event loop
DEBUG 22:09:12: rmp_rpc::endpoint: ServerEndpoint: polling  <------------------------- The endpoint is polled right away
DEBUG 22:09:12: rmp_rpc::endpoint: InnerEndpoint: polling
DEBUG 22:09:12: rmp_rpc::endpoint: Server: flushing responses <----------------------- Nothing to flush yet, because we havent handled any request
TRACE 22:09:12: tokio_io::_tokio_codec::framed_write: flushing framed transport
TRACE 22:09:12: tokio_io::_tokio_codec::framed_write: framed transport flushed
DEBUG 22:09:12: rmp_rpc::endpoint: Polling stream.  <--------------------------------- Polling the TCP stream
DEBUG 22:09:12: rmp_rpc::endpoint: Transport: polling
TRACE 22:09:12: mio::poll: registering with poller
DEBUG 22:09:12: rmp_rpc::endpoint: notifying the reactor that we're not done yet <---- Returning Async::NotReady
TRACE 22:09:12: tokio_reactor: event Readable | Writable Token(4194305)
TRACE 22:09:12: tokio_threadpool::task:     -> not ready
TRACE 22:09:12: tokio_threadpool::notifier: Notifier::notify; id=0x7f0387652080
TRACE 22:09:12: tokio_threadpool::worker: Worker::sleep; worker=WorkerId(2)
TRACE 22:09:12: tokio_threadpool::pool: submit to existing worker; idx=3; state=worker::State { lifecycle: Sleeping, is_pushed: true }
TRACE 22:09:12: tokio_threadpool::worker:   sleeping -- push to stack; idx=2
TRACE 22:09:12: tokio_threadpool::worker:     -> starting to sleep; idx=2
DEBUG 22:09:12: tokio_reactor: loop process - 1 events, 0.000s
TRACE 22:09:12: tokio_threadpool::worker:     -> wakeup; idx=3
TRACE 22:09:12: tokio_threadpool::worker: found work while draining; signal_work
TRACE 22:09:12: tokio_threadpool::pool: signal_work -- wakeup; idx=2
TRACE 22:09:12: tokio_threadpool::task: Task::run; state=Running
TRACE 22:09:12: tokio_threadpool::worker:     -> wakeup; idx=2
DEBUG 22:09:12: rmp_rpc::endpoint: ServerEndpoint: polling <-------------------------- Our endpoint is polled again, probably because the request arrived (But I don't see a ReadReady event above?)
DEBUG 22:09:12: rmp_rpc::endpoint: InnerEndpoint: polling
DEBUG 22:09:12: rmp_rpc::endpoint: Server: flushing responses
TRACE 22:09:12: tokio_threadpool::worker: Worker::sleep; worker=WorkerId(2)
TRACE 22:09:12: tokio_io::_tokio_codec::framed_write: flushing framed transport
TRACE 22:09:12: tokio_threadpool::worker:   sleeping -- push to stack; idx=2
TRACE 22:09:12: tokio_io::_tokio_codec::framed_write: framed transport flushed
TRACE 22:09:12: tokio_threadpool::worker:     -> starting to sleep; idx=2
DEBUG 22:09:12: rmp_rpc::endpoint: Polling stream. <---------------------------------- Polling the TCP stream
DEBUG 22:09:12: rmp_rpc::endpoint: Transport: polling
TRACE 22:09:12: tokio_io::_tokio_codec::framed_read: attempting to decode a frame
TRACE 22:09:12: tokio_io::_tokio_codec::framed_read: frame decoded from buffer <------ Found a request
DEBUG 22:09:12: rmp_rpc::endpoint: spawning a new task <------------------------------ Handling the request
 INFO 22:09:12: server: polling LongComputation <------------------------------------- Start by polling the new task. If it's Ready, there's no need to spawn it on the event loop
DEBUG 22:09:12: rmp_rpc::endpoint: spawning the task on the event loop <-------------- It's not Ready, so spawn it on the event loop
TRACE 22:09:12: tokio_threadpool::sender: execute; count=2
TRACE 22:09:12: tokio_threadpool::pool:     -> submit internal; idx=3
TRACE 22:09:12: tokio_threadpool::pool: signal_work -- wakeup; idx=2
DEBUG 22:09:12: rmp_rpc::endpoint: Transport: polling <------------------------------- Try to poll the transport layer again (we poll until there's nothing more to read)
TRACE 22:09:12: tokio_threadpool::worker:     -> wakeup; idx=2
TRACE 22:09:12: tokio_io::_tokio_codec::framed_read: attempting to decode a frame <--- Attempting to decode a frame, but there's nothing more
TRACE 22:09:12: tokio_threadpool::worker: stole task
DEBUG 22:09:12: rmp_rpc::endpoint: notifying the reactor that we're not done yet <---- Return Async::NotReady
TRACE 22:09:12: tokio_threadpool::task: Task::run; state=Running
TRACE 22:09:12: tokio_threadpool::task:     -> not ready
 INFO 22:09:12: server: polling LongComputation <------------------------------------- Looks like tokio polls the future we just spawned, but it's not ready
TRACE 22:09:12: tokio_threadpool::task:     -> not ready
TRACE 22:09:12: tokio_threadpool::worker: Worker::sleep; worker=WorkerId(3)
TRACE 22:09:12: tokio_threadpool::worker: try_steal_task -- signal_work; self=2; from=3
TRACE 22:09:12: tokio_threadpool::worker:   sleeping -- push to stack; idx=3
TRACE 22:09:12: tokio_threadpool::pool: signal_work -- spawn; idx=1
TRACE 22:09:12: tokio_threadpool::worker:     -> starting to sleep; idx=3
TRACE 22:09:12: tokio_threadpool::worker: Worker::sleep; worker=WorkerId(2)
TRACE 22:09:12: tokio_threadpool::worker:   sleeping -- push to stack; idx=2
TRACE 22:09:12: tokio_threadpool::worker:     -> starting to sleep; idx=2
TRACE 22:09:12: tokio_threadpool::worker: Worker::sleep; worker=WorkerId(1)
TRACE 22:09:12: tokio_threadpool::worker:   sleeping -- push to stack; idx=1
TRACE 22:09:12: tokio_threadpool::worker:     -> starting to sleep; idx=1
```

`t1 = t0 + 2s`: after two seconds, the second client connects and we can see the
corresponding read event:

```
TRACE 22:09:14: tokio_reactor: event Readable Token(0)
DEBUG 22:09:14: tokio_reactor: loop process - 1 events, 0.000s
```

`t2 = t1 + 3s = t0 + 5s`: I would expect a new Server to be spawned right away, but it
look like 3 seconds pass before anything happens. That means that what actually
wakes our tasks up is the first request becoming ready. In this block, two
things are happening:

- the first request is ready, and we send the response
- we handle the new connection (two seconds after the client actually initiated
  the connection!)

```
TRACE 22:09:16: tokio_threadpool::worker:     -> wakeup; idx=3 <--- there a wakeup that seems to sometimes occur a bit earlier, not sure what it means. I've redone the experiment a couple times, and this sometimes occur 2 or 3 seconds before the rest.
TRACE 22:09:17: tokio_threadpool::worker:     -> wakeup; idx=3
TRACE 22:09:17: tokio_threadpool::notifier: Notifier::notify; id=0x7f0386a46080
TRACE 22:09:17: tokio_threadpool::pool:     -> submit internal; idx=3
TRACE 22:09:17: tokio_threadpool::pool: signal_work -- wakeup; idx=1
TRACE 22:09:17: tokio_threadpool::worker:     -> wakeup; idx=3
TRACE 22:09:17: tokio_threadpool::worker:     -> wakeup; idx=1
TRACE 22:09:17: tokio_threadpool::worker: stole task
TRACE 22:09:17: tokio_threadpool::task: Task::run; state=Running
 INFO 22:09:17: server: polling LongComputation <---------------------------------- The future corresponding to the first request is polled
TRACE 22:09:17: tokio_threadpool::notifier: Notifier::notify; id=0x7f0387652080
TRACE 22:09:17: tokio_threadpool::pool:     -> submit internal; idx=1
TRACE 22:09:17: tokio_threadpool::pool: signal_work -- wakeup; idx=2
TRACE 22:09:17: tokio_threadpool::task:     -> task complete
TRACE 22:09:17: tokio_threadpool::worker:     -> wakeup; idx=2
TRACE 22:09:17: tokio_threadpool::worker: task complete; state=pool::State { lifecycle: ShutdownOnIdle, num_futures: 1 } <--- It's Ready!
TRACE 22:09:17: tokio_threadpool::worker: stole task
TRACE 22:09:17: tokio_threadpool::worker: try_steal_task -- signal_work; self=1; from=3
TRACE 22:09:17: tokio_threadpool::task: Task::run; state=Running
TRACE 22:09:17: tokio_threadpool::pool: signal_work -- wakeup; idx=3
DEBUG 22:09:17: rmp_rpc::endpoint: ServerEndpoint: polling <------------------------ That wakes up the task corresponding to the first connection
DEBUG 22:09:17: rmp_rpc::endpoint: InnerEndpoint: polling
TRACE 22:09:17: tokio_threadpool::worker:     -> wakeup; idx=3
DEBUG 22:09:17: rmp_rpc::endpoint: Server: flushing responses <--------------------- We're sending the response back to the client
TRACE 22:09:17: tokio_threadpool::worker: Worker::sleep; worker=WorkerId(1)
TRACE 22:09:17: tokio_threadpool::worker: Worker::sleep; worker=WorkerId(3)
TRACE 22:09:17: tokio_threadpool::worker:   sleeping -- push to stack; idx=1
TRACE 22:09:17: tokio_threadpool::worker:   sleeping -- push to stack; idx=3
TRACE 22:09:17: tokio_threadpool::worker:     -> starting to sleep; idx=1
TRACE 22:09:17: tokio_threadpool::worker:     -> starting to sleep; idx=3
TRACE 22:09:17: tokio_io::_tokio_codec::framed_write: flushing framed transport
TRACE 22:09:17: tokio_io::_tokio_codec::framed_write: writing; remaining=9
TRACE 22:09:17: tokio_io::_tokio_codec::framed_write: framed transport flushed
DEBUG 22:09:17: rmp_rpc::endpoint: Polling stream. <--------------------------------- Polling the TCP stream, in case there are new requests, but there is no new request for this client
DEBUG 22:09:17: rmp_rpc::endpoint: Transport: polling
DEBUG 22:09:17: rmp_rpc::endpoint: notifying the reactor that we're not done yet <--- Returning NotReady
TRACE 22:09:17: tokio_threadpool::task:     -> not ready
TRACE 22:09:17: tokio_threadpool::worker: try_steal_task -- signal_work; self=2; from=1
TRACE 22:09:17: tokio_threadpool::pool: signal_work -- wakeup; idx=3
TRACE 22:09:17: tokio_threadpool::worker:     -> wakeup; idx=3
TRACE 22:09:17: tokio_threadpool::worker: Worker::sleep; worker=WorkerId(2)
TRACE 22:09:17: tokio_threadpool::worker:   sleeping -- push to stack; idx=2
TRACE 22:09:17: tokio_threadpool::worker:     -> starting to sleep; idx=2
TRACE 22:09:17: tokio_threadpool::worker: Worker::sleep; worker=WorkerId(3)
TRACE 22:09:17: tokio_threadpool::worker:   sleeping -- push to stack; idx=3
TRACE 22:09:17: tokio_threadpool::worker:     -> starting to sleep; idx=3
TRACE 22:09:17: tokio_reactor: event Readable | Writable | Hup Token(4194305) <------- Something happens here. Probably the first client disconnecting, because right after that we poll its connection
TRACE 22:09:17: tokio_threadpool::notifier: Notifier::notify; id=0x7f0387652080
TRACE 22:09:17: tokio_threadpool::pool: submit to existing worker; idx=3; state=worker::State { lifecycle: Sleeping, is_pushed: true }
DEBUG 22:09:17: tokio_reactor: loop process - 1 events, 0.000s
TRACE 22:09:17: tokio_threadpool::worker:     -> wakeup; idx=3
TRACE 22:09:17: tokio_threadpool::worker: found work while draining; signal_work
TRACE 22:09:17: tokio_threadpool::pool: signal_work -- wakeup; idx=2
TRACE 22:09:17: tokio_threadpool::task: Task::run; state=Running
TRACE 22:09:17: tokio_threadpool::worker:     -> wakeup; idx=2
DEBUG 22:09:17: rmp_rpc::endpoint: ServerEndpoint: polling <-------------------------- Polling the first client task again
DEBUG 22:09:17: rmp_rpc::endpoint: InnerEndpoint: polling
DEBUG 22:09:17: rmp_rpc::endpoint: Server: flushing responses <----------------------- Flushing the responses (we already sent the response though, so nothing happens)
TRACE 22:09:17: tokio_threadpool::worker: Worker::sleep; worker=WorkerId(2)
TRACE 22:09:17: tokio_io::_tokio_codec::framed_write: flushing framed transport
TRACE 22:09:17: tokio_threadpool::worker:   sleeping -- push to stack; idx=2
TRACE 22:09:17: tokio_io::_tokio_codec::framed_write: framed transport flushed
DEBUG 22:09:17: rmp_rpc::endpoint: Polling stream. <----------------------------------- Polling for new request
TRACE 22:09:17: tokio_threadpool::worker:     -> starting to sleep; idx=2
DEBUG 22:09:17: rmp_rpc::endpoint: Transport: polling
DEBUG 22:09:17: rmp_rpc::endpoint: Stream closed by remote peer. <--------------------- Noticing that the stream is actually closed, meaning that the client disconnected
TRACE 22:09:17: mio::poll: deregistering handle with poller
DEBUG 22:09:17: tokio_reactor: dropping I/O source: 1
 INFO 22:09:17: server: new connection TcpStream { addr: V4(127.0.0.1:54321), peer: V4(127.0.0.1:44696), fd: 7 } <---- Handling the new connection, FINALLY!
 INFO 22:09:17: server: spawning a new Server
DEBUG 22:09:17: rmp_rpc::endpoint: ServerEndpoint: polling <---------------------------- From there, everything is similar to the first connection. First, we poll the endpoint
DEBUG 22:09:17: rmp_rpc::endpoint: InnerEndpoint: polling
DEBUG 22:09:17: rmp_rpc::endpoint: Server: flushing responses
TRACE 22:09:17: tokio_io::_tokio_codec::framed_write: flushing framed transport
TRACE 22:09:17: tokio_io::_tokio_codec::framed_write: framed transport flushed
DEBUG 22:09:17: rmp_rpc::endpoint: Polling stream. <------------------------------------ We poll the stream (there's no request yet)
DEBUG 22:09:17: rmp_rpc::endpoint: Transport: polling
TRACE 22:09:17: mio::poll: registering with poller
DEBUG 22:09:17: rmp_rpc::endpoint: notifying the reactor that we're not done yet <------ Returning NotRead
TRACE 22:09:17: tokio_threadpool::task:     -> not ready
TRACE 22:09:17: tokio_reactor: event Readable | Writable Token(8388609) <--------------- Event upon receiving the request from the second client
TRACE 22:09:17: tokio_threadpool::notifier: Notifier::notify; id=0x7f0387652080
TRACE 22:09:17: tokio_threadpool::pool: submit to existing worker; idx=2; state=worker::State { lifecycle: Sleeping, is_pushed: true }
DEBUG 22:09:17: tokio_reactor: loop process - 1 events, 0.000s
TRACE 22:09:17: tokio_threadpool::worker:     -> wakeup; idx=2
TRACE 22:09:17: tokio_threadpool::worker: found work while draining; signal_work
TRACE 22:09:17: tokio_threadpool::worker: Worker::sleep; worker=WorkerId(3)
TRACE 22:09:17: tokio_threadpool::pool: signal_work -- wakeup; idx=1
TRACE 22:09:17: tokio_threadpool::worker:   sleeping -- push to stack; idx=3
TRACE 22:09:17: tokio_threadpool::task: Task::run; state=Running
TRACE 22:09:17: tokio_threadpool::worker:     -> wakeup; idx=1
DEBUG 22:09:17: rmp_rpc::endpoint: ServerEndpoint: polling <----------------------------- Polling the endpoint again
TRACE 22:09:17: tokio_threadpool::worker:     -> starting to sleep; idx=3
DEBUG 22:09:17: rmp_rpc::endpoint: InnerEndpoint: polling
DEBUG 22:09:17: rmp_rpc::endpoint: Server: flushing responses
TRACE 22:09:17: tokio_threadpool::worker: Worker::sleep; worker=WorkerId(1)
TRACE 22:09:17: tokio_io::_tokio_codec::framed_write: flushing framed transport
TRACE 22:09:17: tokio_threadpool::worker:   sleeping -- push to stack; idx=1
TRACE 22:09:17: tokio_io::_tokio_codec::framed_write: framed transport flushed
TRACE 22:09:17: tokio_threadpool::worker:     -> starting to sleep; idx=1
DEBUG 22:09:17: rmp_rpc::endpoint: Polling stream.
DEBUG 22:09:17: rmp_rpc::endpoint: Transport: polling
TRACE 22:09:17: tokio_io::_tokio_codec::framed_read: attempting to decode a frame
TRACE 22:09:17: tokio_io::_tokio_codec::framed_read: frame decoded from buffer <--------- Found the request
DEBUG 22:09:17: rmp_rpc::endpoint: spawning a new task
 INFO 22:09:17: server: polling LongComputation
DEBUG 22:09:17: rmp_rpc::endpoint: spawning the task on the event loop <----------------- Spawning the future corresponding to the request on the event loop
TRACE 22:09:17: tokio_threadpool::sender: execute; count=2
TRACE 22:09:17: tokio_threadpool::pool:     -> submit internal; idx=2
TRACE 22:09:17: tokio_threadpool::pool: signal_work -- wakeup; idx=1
DEBUG 22:09:17: rmp_rpc::endpoint: Transport: polling
TRACE 22:09:17: tokio_threadpool::worker:     -> wakeup; idx=1
TRACE 22:09:17: tokio_io::_tokio_codec::framed_read: attempting to decode a frame <------ There's no more request
TRACE 22:09:17: tokio_threadpool::worker: stole task
TRACE 22:09:17: tokio_threadpool::task: Task::run; state=Running
 INFO 22:09:17: server: polling LongComputation <---------------------------------------- In the meantime, tokio polls the future we just spawned
TRACE 22:09:17: tokio_threadpool::task:     -> not ready
TRACE 22:09:17: tokio_threadpool::worker: try_steal_task -- signal_work; self=1; from=2
TRACE 22:09:17: tokio_threadpool::pool: signal_work -- wakeup; idx=3
DEBUG 22:09:17: rmp_rpc::endpoint: notifying the reactor that we're not done yet <-------- Returning NotReady
TRACE 22:09:17: tokio_threadpool::worker:     -> wakeup; idx=3
TRACE 22:09:17: tokio_threadpool::task:     -> not ready
TRACE 22:09:17: tokio_threadpool::worker: Worker::sleep; worker=WorkerId(3)
TRACE 22:09:17: tokio_threadpool::worker: Worker::sleep; worker=WorkerId(1)
TRACE 22:09:17: tokio_threadpool::worker:   sleeping -- push to stack; idx=3
TRACE 22:09:17: tokio_threadpool::worker:   sleeping -- push to stack; idx=1
TRACE 22:09:17: tokio_threadpool::worker:     -> starting to sleep; idx=3
TRACE 22:09:17: tokio_threadpool::worker: Worker::sleep; worker=WorkerId(2)
TRACE 22:09:17: tokio_threadpool::worker:   sleeping -- push to stack; idx=2
TRACE 22:09:17: tokio_threadpool::worker:     -> starting to sleep; idx=2
TRACE 22:09:17: tokio_threadpool::worker:     -> starting to sleep; idx=1
```

`t3 = t2 + 5s = t0 + 10s`: The second client's request finishes we send the
response, and the client disconnects. Nothing very interesting here.

```
TRACE 22:09:21: tokio_threadpool::worker:     -> wakeup; idx=2
TRACE 22:09:22: tokio_threadpool::worker:     -> wakeup; idx=2
TRACE 22:09:22: tokio_threadpool::notifier: Notifier::notify; id=0x7f038644e0f0
TRACE 22:09:22: tokio_threadpool::pool:     -> submit internal; idx=2
TRACE 22:09:22: tokio_threadpool::pool: signal_work -- wakeup; idx=2
TRACE 22:09:22: tokio_threadpool::worker:     -> wakeup; idx=2
TRACE 22:09:22: tokio_threadpool::task: Task::run; state=Running
 INFO 22:09:22: server: polling LongComputation
TRACE 22:09:22: tokio_threadpool::notifier: Notifier::notify; id=0x7f0387652080
TRACE 22:09:22: tokio_threadpool::pool:     -> submit internal; idx=2
TRACE 22:09:22: tokio_threadpool::pool: signal_work -- wakeup; idx=1
TRACE 22:09:22: tokio_threadpool::task:     -> task complete
TRACE 22:09:22: tokio_threadpool::worker:     -> wakeup; idx=1
TRACE 22:09:22: tokio_threadpool::worker: task complete; state=pool::State { lifecycle: ShutdownOnIdle, num_futures: 1 }
TRACE 22:09:22: tokio_threadpool::worker: stole task
TRACE 22:09:22: tokio_threadpool::task: Task::run; state=Running
TRACE 22:09:22: tokio_threadpool::worker: Worker::sleep; worker=WorkerId(2)
DEBUG 22:09:22: rmp_rpc::endpoint: ServerEndpoint: polling
TRACE 22:09:22: tokio_threadpool::worker:   sleeping -- push to stack; idx=2
DEBUG 22:09:22: rmp_rpc::endpoint: InnerEndpoint: polling
TRACE 22:09:22: tokio_threadpool::worker:     -> starting to sleep; idx=2
DEBUG 22:09:22: rmp_rpc::endpoint: Server: flushing responses
TRACE 22:09:22: tokio_io::_tokio_codec::framed_write: flushing framed transport
TRACE 22:09:22: tokio_io::_tokio_codec::framed_write: writing; remaining=9
TRACE 22:09:22: tokio_io::_tokio_codec::framed_write: framed transport flushed
DEBUG 22:09:22: rmp_rpc::endpoint: Polling stream.
DEBUG 22:09:22: rmp_rpc::endpoint: Transport: polling
DEBUG 22:09:22: rmp_rpc::endpoint: notifying the reactor that we're not done yet
TRACE 22:09:22: tokio_threadpool::task:     -> not ready
TRACE 22:09:22: tokio_threadpool::worker: try_steal_task -- signal_work; self=1; from=2
TRACE 22:09:22: tokio_threadpool::pool: signal_work -- wakeup; idx=2
TRACE 22:09:22: tokio_threadpool::worker:     -> wakeup; idx=2
TRACE 22:09:22: tokio_threadpool::worker: Worker::sleep; worker=WorkerId(1)
TRACE 22:09:22: tokio_threadpool::worker:   sleeping -- push to stack; idx=1
TRACE 22:09:22: tokio_threadpool::worker:     -> starting to sleep; idx=1
TRACE 22:09:22: tokio_threadpool::worker: Worker::sleep; worker=WorkerId(2)
TRACE 22:09:22: tokio_threadpool::worker:   sleeping -- push to stack; idx=2
TRACE 22:09:22: tokio_threadpool::worker:     -> starting to sleep; idx=2
TRACE 22:09:22: tokio_reactor: event Readable | Writable | Hup Token(8388609)
TRACE 22:09:22: tokio_threadpool::notifier: Notifier::notify; id=0x7f0387652080
TRACE 22:09:22: tokio_threadpool::pool: submit to existing worker; idx=2; state=worker::State { lifecycle: Sleeping, is_pushed: true }
DEBUG 22:09:22: tokio_reactor: loop process - 1 events, 0.000s
TRACE 22:09:22: tokio_threadpool::worker:     -> wakeup; idx=2
TRACE 22:09:22: tokio_threadpool::worker: found work while draining; signal_work
TRACE 22:09:22: tokio_threadpool::pool: signal_work -- wakeup; idx=1
TRACE 22:09:22: tokio_threadpool::task: Task::run; state=Running
DEBUG 22:09:22: rmp_rpc::endpoint: ServerEndpoint: polling
DEBUG 22:09:22: rmp_rpc::endpoint: InnerEndpoint: polling
DEBUG 22:09:22: rmp_rpc::endpoint: Server: flushing responses
TRACE 22:09:22: tokio_io::_tokio_codec::framed_write: flushing framed transport
TRACE 22:09:22: tokio_io::_tokio_codec::framed_write: framed transport flushed
DEBUG 22:09:22: rmp_rpc::endpoint: Polling stream.
DEBUG 22:09:22: rmp_rpc::endpoint: Transport: polling
DEBUG 22:09:22: rmp_rpc::endpoint: Stream closed by remote peer.
TRACE 22:09:22: mio::poll: deregistering handle with poller
DEBUG 22:09:22: tokio_reactor: dropping I/O source: 1
TRACE 22:09:22: tokio_threadpool::task:     -> not ready
TRACE 22:09:22: tokio_threadpool::worker:     -> wakeup; idx=1
TRACE 22:09:22: tokio_threadpool::worker: Worker::sleep; worker=WorkerId(2)
TRACE 22:09:22: tokio_threadpool::worker:   sleeping -- push to stack; idx=2
TRACE 22:09:22: tokio_threadpool::worker:     -> starting to sleep; idx=2
TRACE 22:09:22: tokio_threadpool::worker: Worker::sleep; worker=WorkerId(1)
TRACE 22:09:22: tokio_threadpool::worker:   sleeping -- push to stack; idx=1
TRACE 22:09:22: tokio_threadpool::worker:     -> starting to sleep; idx=1
```

**Comment by little-dude** [comment]
@tylerwhall do you have any idea what could be wrong here? I have to confess I'm a bit puzzled.

**Comment by little-dude** [comment]
Oooohhh I got it @Riamse ! https://github.com/tokio-rs/tokio-core/blob/master/examples/echo.rs#L105-L117

I thought that `for_each` would spawn the future on the event loop but that's not the case! It just processes the future for us! We need to actually _spawn_ the server:

```patch
-            serve(stream, Server).map_err(|e| info!("server error {}", e))               
+            tokio::spawn(serve(stream, Server).map_err(|e| info!("server error {}", e))) 
```

**Comment by ghost** [comment]
Thank you for taking the weekend to work on this.

**Comment by ghost** [comment]
@little-dude  Is there any reason that the new commit will make your example server work, but not the one I tried to do initially?

**Comment by ghost** [comment]
Ah, yes, there is, because I just blindly copied and pasted my code without adding the `tokio::spawn` line. Never mind - case closed, and again, thank you for all the help!

**Comment by little-dude** [comment]
> Thank you for taking the weekend to work on this.

No worries, Rust is a hobby for me :)

>  I just blindly copied and pasted my code without adding the tokio::spawn line.

Yeah I'm glad it was only that, at first I thought something was really really wrong with the library.
